src/FluentValidation/DefaultValidatorOptions.cs

```diff
@@ -352,6 +352,41 @@ public static class DefaultValidatorOptions {
 			return rule.Configure(config => config.CurrentValidator.GetMetadata().Severity = severity);
 		}
 
+		public static IRuleBuilderOptions<T, TProperty> OnFailure<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Action<T> onFailure) {
+			return rule.Configure(config => config.ReplaceValidator(config.CurrentValidator, new OnFailureValidator<T, TProperty>(config.CurrentValidator, onFailure)));
+		}
+
+		/// <summary>
+		/// Specifies custom method that will be called when specific rule fails
+		/// </summary>
+		/// <typeparam name="T"></typeparam>
+		/// <typeparam name="TProperty"></typeparam>
+		/// <param name="rule"></param>
+		/// <param name="onFailure"></param>
+		/// <returns></returns>
+		public static IRuleBuilderOptions<T, TProperty> OnFailure<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Action<T, PropertyValidatorContext> onFailure) {
+			return rule.Configure(config => config.ReplaceValidator(config.CurrentValidator, new OnFailureValidator<T, TProperty>(config.CurrentValidator, onFailure)));

```

Comment:
Wrapping the validator using `ReplaceValidator` is fine, but it does mean that it breaks the TestHelper (specifically `ShouldHaveChildValidator`), so this PR will need to address that too.

Reply:
Can I achieve the same result without wrapping validator? If Yes then I can do that to not break anything else, if Not then what should I test for?

Reply:
Without a wrapper you'd have to make changes to PropertyValidator which I wouldn't recommend. 

Stick with the wrapper, but basically you'd need to test that this works:

```
_validator.RuleFor(person => person.Address).SetValidator(new AddressValidator().OnFailure((person, ctx) => { });

// this should work:
_validator.ShouldHaveChildValidator(x => x.Address, typeof(AddressValidator));
// but it won't, because it'll find an instance of OnFailureValidator instead
```

And then the fix will be here: https://github.com/JeremySkinner/FluentValidation/blob/248682d7ef902eab48ccd5c9bb679e1d8026abd4/src/FluentValidation/TestHelper/ValidatorTestExtensions.cs#L74

That has a special case for unwrapping the validator if it's a ChildValidatorAdaptor. You'll need to add the OnFailureValidator here. It might actually be nice to add an interface for this, but I'm happy for you to hard code the type check for OnFailureValidator in there for now.